### PR TITLE
collectors/GoCollector: Add tests and examples

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

### CODE REVIEW

File: prometheus/collectors/go_collector_latest_test.go [hunk 1]
```diff
@@ -40,9 +44,331 @@ func TestGoCollectorMarshalling(t *testing.T) {
 		t.Errorf("json marshalling shoud not fail, %v", err)
 	}
 }
+func TestWithGoCollectorMemStatsMetricsDisabled(t *testing.T) {
```
Comment (bwplotka) [inline prometheus/collectors/go_collector_latest_test.go:47]: I think we don't need that. This is thoroughly tested in https://github.com/prometheus/client_golang/blob/b94d7e25292465ac578c2e885e6d61e03a81ec4b/prometheus/go_collector_latest_test.go#L77

File: prometheus/collectors/go_collector_latest_test.go [hunk 2]
```diff
@@ -40,9 +44,331 @@ func TestGoCollectorMarshalling(t *testing.T) {
...
+	for _, r := range result {
+		got = append(got, r.GetName())
+	}
+
+	if !reflect.DeepEqual(got, expected) {
+		t.Errorf("got %v, want %v", got, expected)
+	}
+}
+
+func TestGoCollectorAllowList(t *testing.T) {
```
Comment (bwplotka) [inline prometheus/collectors/go_collector_latest_test.go:74]: Same here: https://github.com/prometheus/client_golang/blob/b94d7e25292465ac578c2e885e6d61e03a81ec4b/prometheus/go_collector_latest_test.go#L77

File: prometheus/collectors/go_collector_latest_test.go [hunk 3]
```diff
@@ -40,9 +44,331 @@ func TestGoCollectorMarshalling(t *testing.T) {
...
+			}
+
+			if !reflect.DeepEqual(got, test.expected) {
+				t.Errorf("got %v, want %v", got, test.expected)
+			}
+		})
+	}
+}
+
+func TestGoCollectorDenyList(t *testing.T) {
```
Comment (bwplotka) [inline prometheus/collectors/go_collector_latest_test.go:218]: Ditto https://github.com/prometheus/client_golang/blob/b94d7e25292465ac578c2e885e6d61e03a81ec4b/prometheus/go_collector_latest_test.go#L77

File: prometheus/collectors/go_collector_latest_test.go [hunk 4]
```diff
@@ -40,9 +44,331 @@ func TestGoCollectorMarshalling(t *testing.T) {
...
 func ExampleGoCollector_WithAdvancedGoMetrics() {
+	reg := prometheus.NewRegistry()
+
+	reg.MustRegister(
+		NewGoCollector(
+			WithGoCollectorRuntimeMetrics(
+				MetricsScheduler,
+				MetricsGC,
+				GoRuntimeMetricsRule{
+					Matcher: regexp.MustCompile("^/mycustomrule.*"),
```
Comment (bwplotka) [inline prometheus/collectors/go_collector_latest_test.go:367]: I would use `gc` example and comment what metrics you will see (:

**Review by bwplotka — COMMENTED** [review]
Nice! Some comments (:

File: prometheus/collectors/go_collector_latest_test.go [hunk 5]
```diff
@@ -40,9 +44,331 @@ func TestGoCollectorMarshalling(t *testing.T) {
...
 func ExampleGoCollector() {
+	reg := prometheus.NewRegistry()
+
+	reg.MustRegister(NewGoCollector())
+
+	http.Handle("/metrics", promhttp.HandlerFor(reg, promhttp.HandlerOpts{}))
+	log.Fatal(http.ListenAndServe(":8080", nil))
 }
 
 func ExampleGoCollector_WithAdvancedGoMetrics() {
```
Comment (bwplotka) [inline prometheus/collectors/go_collector_latest_test.go:358]: We can also add example for helpers you have prepared 👍🏽

File: prometheus/collectors/go_collector_latest.go [hunk 6]
```diff
@@ -90,6 +97,7 @@ func WithGoCollectorRuntimeMetrics(rules ...GoRuntimeMetricsRule) func(options *
 
 	return func(o *internal.GoCollectorOptions) {
 		o.DisableMemStatsLikeMetrics = true
+		o.RuntimeMetricRules = append(o.RuntimeMetricRules, rs...)
```
Comment (kakkoyun) [inline prometheus/collectors/go_collector_latest.go:100]: @bwplotka

File: prometheus/collectors/go_collector_latest_test.go [hunk 7]
```diff
@@ -40,9 +44,331 @@ func TestGoCollectorMarshalling(t *testing.T) {
...
+				"go_memstats_stack_sys_bytes",
+				"go_memstats_sys_bytes",
+				"go_threads",
+			},
+		},
+		{
+			name:     "deny all",
+			matchers: []*regexp.Regexp{regexp.MustCompile("/.*")},
+			expected: []string{
+				"go_gc_duration_seconds",
```
Comment (kakkoyun) [inline prometheus/collectors/go_collector_latest_test.go:261]: @bwplotka

File: prometheus/collectors/go_collector_latest_test.go (hunk 4 above)
Comment (bwplotka) [reply prometheus/collectors/go_collector_latest_test.go:367]: ah actually I missed you already added

File: prometheus/collectors/go_collector_latest_test.go (hunk 2 above)
Comment (kakkoyun) [reply prometheus/collectors/go_collector_latest_test.go:77]: Let's have these as examples for visual people like me.

File: prometheus/collectors/go_collector_latest_test.go (hunk 3 above)
Comment (kakkoyun) [reply prometheus/collectors/go_collector_latest_test.go:207]: Let's have these as examples for visual people like me.

File: prometheus/collectors/go_collector_latest.go [hunk 8]
```diff
@@ -23,7 +23,14 @@ import (
 	"github.com/prometheus/client_golang/prometheus/internal"
 )
 
-// WithGoCollectorMemStatsMetricsDisabled metrics that where representing runtime.MemStats structure such as:
+var (
+	MetricsAll       = regexp.MustCompile("/.*")
```
Comment (bwplotka) [inline prometheus/collectors/go_collector_latest.go:27]: Perhaps we need comments on those 🙈

**Review by bwplotka — APPROVED** [review]
LGTM, just small nit (:

File: prometheus/collectors/go_collector_latest.go (hunk 8 above)
Comment (kakkoyun) [reply prometheus/collectors/go_collector_latest.go:27]: So true :)

**Review by bwplotka — APPROVED** [review]
👍🏽

**Pull request merged**